pkg/fqdn/dnsproxy/proxy.go

```diff
@@ -759,16 +765,141 @@ func (p *DNSProxy) CheckAllowed(endpointID uint64, destPort uint16, destID ident
 	return false, nil
 }
 
-func setSoMark(fd int, secId identity.NumericIdentity) error {
-	mark := linux_defaults.MagicMarkIdentity
+func setSoMarks(fd int, ipFamily ipfamily.IPFamily, secId identity.NumericIdentity) error {
+	if err := unix.SetsockoptInt(fd, ipFamily.SocketOptsFamily, ipFamily.SocketOptsTransparent, 1); err != nil {
+		return fmt.Errorf("setsockopt(IP_TRANSPARENT) for %s failed: %w", ipFamily.Name, err)
+	}
+
+	mark := linux_defaults.MagicMarkEgress
 	mark |= int(uint32(secId&0xFFFF)<<16 | uint32((secId&0xFF0000)>>16))
 	err := unix.SetsockoptInt(fd, unix.SOL_SOCKET, unix.SO_MARK, mark)
 	if err != nil {
 		return fmt.Errorf("error setting SO_MARK: %w", err)
 	}
+
+	if err := unix.SetsockoptInt(int(fd), unix.SOL_SOCKET, unix.SO_REUSEADDR, 1); err != nil {
+		return fmt.Errorf("setsockopt(SO_REUSEADDR) failed: %w", err)
+	}
+	if !option.Config.EnableBPFTProxy {
+		if err := unix.SetsockoptInt(int(fd), unix.SOL_SOCKET, unix.SO_REUSEPORT, 1); err != nil {
+			return fmt.Errorf("setsockopt(SO_REUSEPORT) failed: %w", err)
+		}
+	}
+
 	return nil
 }
 
+var errNoReader = errors.New("Reader stopped")
+
+type Response struct {
+	*dns.Msg
+	err error
+}
+
+// A Client keeps state for concurrent transactions on the same upstream connection.
+type Client struct {
+	dns.Client
+	refcount int // protected by proxy's lock
+
+	// Client's connection shared among all requests from the same source address/port. The
+	// locks below is used to serialize reads and writes on this connection, but reads and
+	// writes can happen at the same time.
+	conn *dns.Conn
+
+	// this mutex protects writes on 'conn' and all access to 'reqs'
+	lock.RWMutex
+	reqs map[uint16]chan Response // outstanding requests
+
+	// 'reader' mutex is used to serialize reads on 'conn'. It is always taken and released
+	// while holding the lock but the lock can be released and re-acquired while holding
+	// 'reader' mutex.
+	reader lock.Mutex
+}
+
+// ExchangeAsync writes the request to the Client's connection and co-operatively
+// reads responses from the connection and distributes them to the requestors.
+// At most one caller is reading from Client's connection at any time.
+func (c *Client) ExchangeAsync(m *dns.Msg) (r *dns.Msg, rtt time.Duration, err error) {
+	// Lock allows only one request to be written at a time, but that can happen
+	// concurrently with reading.
+	c.Lock()
+	defer c.Unlock()
+	if _, exists := c.reqs[m.Id]; exists {
+		return nil, 0, fmt.Errorf("Duplicate request: %d", m.Id)
```

Comment:
nit: Go errors should be lower case
```suggestion
		return nil, 0, fmt.Errorf("duplicate request: %d", m.Id)
```

Reply:
fixed, thanks!